[JIT] SubgraphMatcher: add attributes support.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20602 [JIT] SubgraphMatcher: add attributes support.**

Differential Revision: [D15377635](https://our.internmc.facebook.com/intern/diff/D15377635)

Commit:
[JIT] SubgraphMatcher: add attributes support.

nishantpdce commented on test/cpp/jit/test_subgraph_matcher.h:
```diff
@@ -361,6 +361,80 @@ graph(%x, %y):
+  script::parseIR(
+      R"IR(
+graph(%0):
+  %a = a::a[isattr=[1,2]](%0)
```
Looks mostly good. Question - for the case we are interested in, we will be matching only particular attribute [4th arg of dequantize_linear which is static], rest all are dynamic. How will the pattern look like for that case ?

ZolotukhinM replied:
Good question, I don't have a definite answer to that: we could either 1) relax the check a little and verify that only attributes that are present match, or 2) introduce wildcard support for attributes (it's more sophisticated because we need to come up with a sufficiently general and yet not too complicated to use design here).

But for our use case, do you expect *attributes* to be dynamic? I though we would just have one argument that would be const - for that we would have a node in the pattern with the attribute, and this solution should work there - other arguments would just match as they did before (without any attributes). Am I missing something?

zdevito commented on torch/csrc/jit/subgraph_matcher.cpp:
```diff
@@ -140,6 +141,33 @@ bool SubgraphMatcher::matchNodes(const Node* n1, Node* n2) {
+        }
+        break;
+      default:
+        // Other attributes types not supported yet
```
Should this error? Or is this not reachable?

ZolotukhinM replied:
If we see an attribute of unknown type, we just conservatively don't match. It is reachable, but I'm not sure if we want to hard fail here.

Commit:
Clang-tidy on "[JIT] SubgraphMatcher: add attributes support."

[JIT] SubgraphMatcher: add attributes support.

gh-metadata: pytorch pytorch 20602 gh/ZolotukhinM/89/head
